src/wiring_diagrams/ScheduleUndirected.jl

```diff
@@ -0,0 +1,219 @@
+
+  parent::Hom(Composite, Composite)
+  box_parent::Hom(Box, Composite)
+
+  # Po-category axiom enforcing that composites form a rooted forest.
+  # parent <= id(Composite)
+end
+
+const AbstractScheduledUWD = AbstractACSetType(TheoryScheduledUWD)
+
+""" Scheduled undirected wiring diagram.
+
+A scheduled UWD consists of a UWD together with a set of *composite* nodes
+forming a rooted tree, or in general a rooted forest, whose leaves are the
+diagram's boxes.
+
+See also: [`NestedUWD`](@ref).
+"""
+const ScheduledUWD = CSetType(TheoryScheduledUWD,
+  index=[:box, :junction, :outer_junction, :parent, :box_parent])
+
+ncomposites(x::AbstractACSet) = nparts(x, :Composite)
+composites(x::AbstractACSet) = parts(x, :Composite)
+parent(x::AbstractACSet, args...) = subpart(x, args..., :parent)
+children(x::AbstractACSet, c::Int) =
+  filter(c′ -> c′ != c, incident(x, c, :parent))
+box_parent(x::AbstractACSet, args...) = subpart(x, args..., :box_parent)
+box_children(x::AbstractACSet, args...) = incident(x, args..., :box_parent)
+
+@present TheoryNestedUWD <: TheoryScheduledUWD begin
+  CompositePort::Ob
+
+  composite::Hom(CompositePort, Composite)
+  composite_junction::Hom(CompositePort, Junction)
+end
+
+const AbstractNestedUWD = AbstractACSetType(TheoryNestedUWD)
+
+""" Nested undirected wiring diagram.
+
+A nested UWD is a scheduled UWD whose composite nodes have been given ports,
+making explicit the intermediate boxes in the composition.
+
+Nested UWDs are very similar but not quite identical to Robin Milner's
+[bigraphs](https://doi.org/10.1017/CBO9780511626661).
+
+See also: [`ScheduledUWD`](@ref).
+"""
+const NestedUWD = CSetType(TheoryNestedUWD,
+  index=[:box, :junction, :outer_junction,
+         :composite, :composite_junction, :parent, :box_parent])
+
+composite_ports(x::AbstractACSet, args...) = incident(x, args..., :composite)
+composite_junction(x::AbstractACSet, args...) =
+  subpart(x, args..., :composite_junction)
+composite_ports_with_junction(x::AbstractACSet, args...) =
+  incident(x, args..., :composite_junction)
+
+# Evaluation
+############
+
+""" Evaluate a scheduled UWD given a set of generators for the boxes.
+
+The first argument `f` should be a callabe supporting the signature
+
+```
+f(values::AbstractVector,
+  juncs::AbstractVector{<:FinFunction{Int}},
+  outer_junc::FinFunction{Int})
+```
+
+where `values` is a list of computed values for boxes or composites in the
+diagram; `juncs` is an equal-length list of `FinSet` functions, mapping the
+ports of each value to junctions; and `outer_junc` is a `FinSet` function
+mapping the outer ports to junctions. This interface is similar to the `ncon`
+function in the [NCON package](https://arxiv.org/abs/1402.0939) for MATLAB or
+the `ncon` function in
+[TensorOperations.jl](https://github.com/Jutho/TensorOperations.jl), except that
+the outer junctions are represented explictly by a third argument rather than
+implictly by using negative numbers in the second argument.
```

Comment:
I don't think that most users of UWD will be familiar with TensorOperations.jl so this should probably be moved down in the docstring?